File: SMT/Models/DiagnosticoPsicologico.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace SMT.Models.DB
{
    public partial class DiagnosticoPsicologico
    {

        public void crear()
        {
            using(SMTDevEntities db = new SMTDevEntities())
            {
                FechaActualizacion = DateTime.Now;
                db.DiagnosticoPsicologico.Add(this);
                db.SaveChanges();
            }

        }

        public void editar()
        {
            using (SMTDevEntities db = new SMTDevEntities())
            {
                DiagnosticoPsicologico diag = db.DiagnosticoPsicologico.FirstOrDefault(i => i.IDAlumno == IDAlumno && i.IDGrupo == IDGrupo);

                if(diag != null)
                {
                    diag.Aburrimiento = Aburrimiento;
                    diag.AutonomiaVerguenza = AutonomiaVerguenza;
                    diag.ComoSeTranslada = ComoSeTranslada;
                    diag.ConductaInadecuada = ConductaInadecuada;
                    diag.ConfianzaDesconfianza = ConfianzaDesconfianza;
                    diag.CTE10 = CTE10;
                    diag.CTE8 = CTE8;
                    diag.CTE5 = CTE5;
                    diag.CuentaConCasa = CuentaConCasa;
                    diag.Edad = Edad;
                    diag.EscolaridadPadres = EscolaridadPadres;
                    diag.EtapaPreconvencional = EtapaPreconvencional;
                    diag.EtapaSesioromotora = EtapaSesioromotora;
                    diag.EvaluacionBloque1 = EvaluacionBloque1;
                    diag.EvaluacionBloque2 = EvaluacionBloque2;
                    diag.EvaluacionBloque3 = EvaluacionBloque3;
                    diag.EvaluacionBloque4 = EvaluacionBloque4;
                    diag.EvaluacionBloque5 = EvaluacionBloque5;
                    diag.ExpresaEmociones = ExpresaEmociones;
                    diag.FamiliaCompleta = FamiliaCompleta;
                    diag.FamiliaMonoparental = FamiliaMonoparental;
                    diag.FechaActualizacion = DateTime.Now;
                    diag.FechaSync = DateTime.Now;
                    diag.Generabilidad = Generabilidad;
                    diag.Hermanos = Hermanos;
                    diag.IdentidadConfusion = IdentidadConfusion;
                    diag.Indiferencia = Indiferencia;
                    diag.IniciativaCulpa = IniciativaCulpa;
                    diag.Integridad = Integridad;
                    diag.Interpersonal = Interpersonal;
                    diag.IntimidadAislamiento = IntimidadAislamiento;
                    diag.Intrapersonal = Intrapersonal;
                    diag.KinestesicaCorporal = KinestesicaCorporal;
                    diag.Laboriosidad = Laboriosidad;
                    diag.Linguistica = Linguistica;
                    diag.LogicoMatematica = LogicoMatematica;
                    diag.LugarFamilia = LugarFamilia;
                    diag.MoralidadConvencional = MoralidadConvencional;
                    diag.MoralidadPreconvencional = MoralidadPreconvencional;
                    diag.MoralidadPrincipiosAutonomos = MoralidadPrincipiosAutonomos;
                    diag.Motivado = Motivado;
                    diag.MuestraIniciativa = MuestraIniciativa;
                    diag.MuestraInteresHijo = MuestraInteresHijo;
                    diag.Musical = Musical;
                    diag.Necesidades = Necesidades;
                    diag.NEE = NEE;
                    diag.OperacionesContretas = OperacionesContretas;
                    diag.OperacionesFormales = OperacionesFormales;
                    diag.ParticipaActividadesEscuela = ParticipaActividadesEscuela;
                    diag.ProgramaBloque1 = ProgramaBloque1;
                    diag.ProgramaBloque2 = ProgramaBloque2;
                    diag.ProgramaBloque3 = ProgramaBloque3;
                    diag.ProgramaBloque4 = ProgramaBloque4;
                    diag.ProgramaBloque5 = ProgramaBloque5;
                    diag.ReconoceValia = ReconoceValia;
                    diag.Regular = Regular;
                    diag.RequiereApoyo = RequiereApoyo;
                    diag.RespetaAsiMismo = RespetaAsiMismo;
                    diag.RespetaReglas = RespetaReglas;
                    diag.ResuelveConfilctos = ResuelveConfilctos;
                    diag.ServicioMedico = ServicioMedico;
                    diag.Sexo = Sexo;
                    diag.SituacionRiesgo = SituacionRiesgo;
                    diag.Talentos = Talentos;
                    diag.Temeroso = Temeroso;
                    diag.TieneComputadora = TieneComputadora;
                    diag.VisualEspacial = VisualEspacial;
                    diag.ViveOtros = ViveOtros;
                    
                    

                    db.SaveChanges();
                }
                else
                {
                    crear();
                }

               
            }

        }

        public static List<DiagnosticoPsicologicoSimple> listar(Guid grupo)
        {
            using(SMTDevEntities db = new SMTDevEntities())
            {
                List<DiagnosticoPsicologicoSimple> result = db.DiagnosticoPsicologico
                    .Where(i => i.IDGrupo == grupo)
                    .Select(i => new DiagnosticoPsicologicoSimple()
                    {
                        Alumno = i.IDAlumno,
                        Aburrimiento = i.Aburrimiento,
                        AutonomiaVerguenza = i.AutonomiaVerguenza,
                        ComoSeTranslada = i.ComoSeTranslada,
                        ConductaInadecuada = i.ConductaInadecuada,
                        ConfianzaDesconfianza = i.ConfianzaDesconfianza,
                        CTE10 = i.CTE10,
                        CTE8 = i.CTE8,
                        CTE5 = i.CTE5,
                        CuentaConCasa = i.CuentaConCasa,
                        Edad = i.Edad,
                        EscolaridadPadres = i.EscolaridadPadres,
                        EtapaPreconvencional = i.EtapaPreconvencional,
                        EtapaSesioromotora = i.EtapaSesioromotora,
                        EvaluacionBloque1 = i.EvaluacionBloque1,
                        EvaluacionBloque2 = i.EvaluacionBloque2,
                        EvaluacionBloque3 = i.EvaluacionBloque3,
                        EvaluacionBloque4 = i.EvaluacionBloque4,
                        EvaluacionBloque5 = i.EvaluacionBloque5,
                        ExpresaEmociones = i.ExpresaEmociones,
                        FamiliaCompleta = i.FamiliaCompleta,
                        FamiliaMonoparental = i.FamiliaMonoparental,
                        FechaActualizacion = i.FechaActualizacion ?? DateTime.Now,
                        Generabilidad = i.Generabilidad,
                        Hermanos = i.Hermanos,
                        IdentidadConfusion = i.IdentidadConfusion,
                        Indiferencia = i.Indiferencia,
                        IniciativaCulpa = i.IniciativaCulpa,
                        Integridad = i.Integridad,
                        Interpersonal = i.Interpersonal,
                        IntimidadAislamiento = i.IntimidadAislamiento,
                        Intrapersonal = i.Intrapersonal,
                        KinestesicaCorporal = i.KinestesicaCorporal,
                        Laboriosidad = i.Laboriosidad,
                        Linguistica = i.Linguistica,
                        LogicoMatematica = i.LogicoMatematica,
                        LugarFamilia = i.LugarFamilia,
                        MoralidadConvencional = i.MoralidadConvencional,
                        MoralidadPreconvencional = i.MoralidadPreconvencional,
                        MoralidadPrincipiosAutonomos = i.MoralidadPrincipiosAutonomos,
                        Motivado = i.Motivado,
                        MuestraIniciativa = i.MuestraIniciativa,
                        MuestraInteresHijo = i.MuestraInteresHijo,
                        Musical = i.Musical,
                        Necesidades = i.Necesidades,
                        NEE = i.NEE,
                        OperacionesContretas = i.OperacionesContretas,
                        OperacionesFormales = i.OperacionesFormales,
                        ParticipaActividadesEscuela = i.ParticipaActividadesEscuela,
                        ProgramaBloque1 = i.ProgramaBloque1,
                        ProgramaBloque2 = i.ProgramaBloque2,
                        ProgramaBloque3 = i.ProgramaBloque3,
                        ProgramaBloque4 = i.ProgramaBloque4,
                        ProgramaBloque5 = i.ProgramaBloque5,
                        ReconoceValia = i.ReconoceValia,
                        Regular = i.Regular,
                        RequiereApoyo = i.RequiereApoyo,
                        RespetaAsiMismo = i.RespetaAsiMismo,
                        RespetaReglas = i.RespetaReglas,
                        ResuelveConfilctos = i.ResuelveConfilctos,
                        ServicioMedico = i.ServicioMedico,
                        Sexo = i.Sexo,
                        SituacionRiesgo = i.SituacionRiesgo,
                        Talentos = i.Talentos,
                        Temeroso = i.Temeroso,
                        TieneComputadora = i.TieneComputadora,
                        VisualEspacial = i.VisualEspacial,
                        ViveOtros = i.ViveOtros
            }).ToList();



                return result;
            }

        }


        public class DiagnosticoPsicologicoSimple
        {
            public Guid Alumno { get; set; }
            public string Aburrimiento { get; set; }
            public string ConductaInadecuada { get; set; }
            public int? Edad { get; set; }
            public string Indiferencia { get; set; }
            public string Interpersonal { get; set; }
            public string Intrapersonal { get; set; }
            public string KinestesicaCorporal { get; set; }
            public string Linguistica { get; set; }
            public string LogicoMatematica { get; set; }

            public string Motivado { get; set; }
            public string MuestraIniciativa { get; set; }
            public string Musical { get; set; }
            public string Necesidades { get; set; }
            public string NEE { get; set; }
            public string Regular { get; set; }
            public string RequiereApoyo { get; set; }
            public string Sexo { get; set; }
            public string SituacionRiesgo { get; set; }
            public string Talentos { get; set; }
            public string Temeroso { get; set; }
            public string VisualEspacial { get; set; }
            public string CTE10 { get; internal set; }
            public string CTE8 { get; internal set; }
            public string AutonomiaVerguenza { get; internal set; }
            public string ComoSeTranslada { get; internal set; }
            public string ConfianzaDesconfianza { get; internal set; }
            public string CTE5 { get; internal set; }
            public string CuentaConCasa { get; internal set; }
            public string EscolaridadPadres { get; internal set; }
            public string EtapaPreconvencional { get; internal set; }
            public string EtapaSesioromotora { get; internal set; }
            public string EvaluacionBloque1 { get; internal set; }
            public string EvaluacionBloque2 { get; internal set; }
            public string EvaluacionBloque3 { get; internal set; }
            public string EvaluacionBloque4 { get; internal set; }
            public string EvaluacionBloque5 { get; internal set; }
            public string ExpresaEmociones { get; internal set; }
            public string FamiliaCompleta { get; internal set; }
            public string FamiliaMonoparental { get; internal set; }
            public DateTime FechaActualizacion { get; internal set; }
            public string Generabilidad { get; internal set; }
            public int? Hermanos { get; internal set; }
            public string IdentidadConfusion { get; internal set; }
            public string IniciativaCulpa { get; internal set; }
            public string Integridad { get; internal set; }
            public string IntimidadAislamiento { get; internal set; }
            public string Laboriosidad { get; internal set; }
            public string LugarFamilia { get; internal set; }
            public string MoralidadConvencional { get; internal set; }
            public string MoralidadPreconvencional { get; internal set; }
            public string MoralidadPrincipiosAutonomos { get; internal set; }
            public string MuestraInteresHijo { get; internal set; }
            public string OperacionesContretas { get; internal set; }
            public string OperacionesFormales { get; internal set; }
            public string ParticipaActividadesEscuela { get; internal set; }
            public string ProgramaBloque1 { get; internal set; }
            public string ProgramaBloque2 { get; internal set; }
            public string ProgramaBloque3 { get; internal set; }
            public string ProgramaBloque4 { get; internal set; }
            public string ProgramaBloque5 { get; internal set; }
            public string ReconoceValia { get; internal set; }
            public string RespetaAsiMismo { get; internal set; }
            public string RespetaReglas { get; internal set; }
            public string ResuelveConfilctos { get; internal set; }
            public string ServicioMedico { get; internal set; }
            public string TieneComputadora { get; internal set; }
            public string ViveOtros { get; internal set; }
        }
    }
}

File: SMT/Scripts/apps/DiagnosticoPsicologico/DiagnosticoPsicologico.js
var DiagnosticoPsicologico = new function () {

    var _datos = [];

    this.desplegarTabla = function (grupo) {

        var content = $('[data-psicologico="' + grupo + '"]').html('<h2 class="text-center"><span class="fa fa-refresh fa-spin"></span></h2>');

       
        Alumnos.listar(grupo).then(function (alumnos) {
            cargarLista(grupo).then(function (diagnosticos) {
                    

                content.html(Templates.items['tablaPsicologico']);


                alumnos.map(function (a) {

                    a.Aburrimiento = '';
                    a.AutonomiaVerguenza = '';
                    a.ComoSeTranslada = '';
                    a.ConductaInadecuada = '';
                    a.ConfianzaDesconfianza = '';
                    a.CTE10 = '';
                    a.CTE5 = '';
                    a.CTE8 = '';
                    a.CuentaConCasa = '';
                    a.Edad = '';
                    a.EscolaridadPadres = '';
                    a.EtapaPreconvencional = '';
                    a.EtapaSesioromotora = '';
                    a.EvaluacionBloque1 = '';
                    a.EvaluacionBloque2 = '';
                    a.EvaluacionBloque3 = '';
                    a.EvaluacionBloque4 = '';
                    a.EvaluacionBloque5 = '';
                    a.ExpresaEmociones = '';
                    a.FamiliaCompleta = '';
                    a.FamiliaMonoparental = '';
                    a.Generabilidad = '';
                    a.Hermanos = '';
                    a.IdentidadConfusion = '';
                    a.Indiferencia = '';
                    a.IniciativaCulpa = '';
                    a.Integridad = '';
                    a.Interpersonal = '';
                    a.IntimidadAislamiento = '';
                    a.Intrapersonal = '';
                    a.KinestesicaCorporal = '';
                    a.Laboriosidad = '';
                    a.Linguistica = '';
                    a.LogicoMatematica = '';
                    a.LugarFamilia = '';
                    a.MoralidadConvencional = '';
                    a.MoralidadPreconvencional = '';
                    a.MoralidadPrincipiosAutonomos = '';
                    a.Motivado = '';
                    a.MuestraIniciativa = '';
                    a.MuestraInteresHijo = '';
                    a.Musical = '';
                    a.Necesidades = '';
                    a.NEE = '';
                    a.OperacionesContretas = '';
                    a.OperacionesFormales = '';
                    a.ParticipaActividadesEscuela = '';
                    a.ProgramaBloque1 = '';
                    a.ProgramaBloque2 = '';
                    a.ProgramaBloque3 = '';
                    a.ProgramaBloque4 = '';
                    a.ProgramaBloque5 = '';
                    a.ReconoceValia = '';
                    a.Regular = '';
                    a.RequiereApoyo = '';
                    a.RespetaAsiMismo = '';
                    a.RespetaReglas = '';
                    a.ResuelveConfilctos = '';
                    a.ServicioMedico = '';
                    a.Sexo = '';
                    a.SituacionRiesgo = '';
                    a.Talentos = '';
                    a.Temeroso = '';
                    a.TieneComputadora = '';
                    a.VisualEspacial = '';
                    a.ViveOtros = '';

                    for (var d in diagnosticos) {
                        var diag = diagnosticos[d];
                        if (a.id == diag.Alumno) {
                            diag.id = diag.Alumno;
                            diag.nombre = a.nombre + ' ' + a.apellidoPaterno + ' ' + a.apellidoMaterno;
                            actualizarObjeto(a,diag);
                            break;
                        }
                    }
                });

                alumnos.map(function (alumno) {
                    content.find('tbody').append(Templates.items['rowPsicologico'].format(alumno));
                });


            });
        });
        

    

    }

    var actualizarObjeto = function (a,b) {
        for (var name in a) {
            if (b[name] != undefined && b[name] != null) {
                a[name] = b[name];
            }
            else {
                a[name] = '';
            }
        }
    }

    var cargarLista = function (grupo) {
        return new Promise(function (success) {

            var data = Cache.validarCache('psico' + grupo);

            if (data == false) {

                $.ajax({
                    url: '/DiagnosticosPsicologicos/listar',
                    type: 'get',
                    data: { grupo: grupo },
                    success: function (data) {
                        _datos = data;
                        Cache.almacenarCache(data, 'psico' + grupo, 1);
                        success(data);
                    }
                });
            }
            else {
                _datos = data;
                success(data);
            }
        });
       


    }

    this.Imprimir = function () {
        open('/DiagnosticosPsicologicos/imprimir?grupo=' + _grupo);
    }

    $('body:not(.visualizando)').delegate('td[data-campo]:not(.selected)', 'focusin', function () {
        $(this).addClass('selected');
        var valor = $(this).text();

        var respuesta = $(this).attr('data-respuesta');
        if (parseInt(respuesta) >= 1 && parseInt(respuesta) <= 11) {
            $(this).html(Templates.items['respuesta' + respuesta]);
            $(this).find('input,select').val(valor);
        }
        else{
            $(this).html('<input type="text" value="' + valor + '" class="from-control fom-control-oculto" />')
        }

        $(this).find('select').focus(function () {
            // Esto hace que se carge el select abierto.. no encontre otra manera
            $(this).attr('size', $(this).attr("expandto"));
        });
        $(this).find('input,select').focus();
    });

    $('body:not(.visualizando)').delegate('td[data-campo] input,td[data-campo] select', 'focusout', function () {
        $(this).parent().removeClass('selected');
        $(this).parent().html($(this).val());
        
    });

    $('body:not(.visualizando)').delegate('tr', 'focusout', function () {
        var tr = $(this);
        setTimeout(function () {
            if (tr.find('input').length == 0) {
                var params = [];

                params.push({
                    name: tr.data('campo'),
                    value: tr.data('id')
                });

                params.push({
                    name: 'IDGrupo',
                    value: _grupo
                });

                tr.find('[data-campo]').each(function () {

                    params.push({
                        name: $(this).data('campo'),
                        value: $(this).text()
                    });
                });

                tr.removeClass('success');
                $.ajax({
                    url: '/DiagnosticosPsicologicos/Actualizar',
                    type: 'post',
                    data: params,
                    beforeSend: function () {
                        tr.resaltar('warning',1500);
                    },
                    success: function (response) {
                        if (response.result == true) {
                            tr.resaltar('info',1000);
                            Cache.vaciar();
                        }
                        else {
                            tr.resaltar('danger',2000);
                        }
                    }
                });
            }
        },200);
        
    });

    Templates.load('tablaPsicologico', '/Scripts/apps/DiagnosticoPsicologico/views/Table.html');
    Templates.load('rowPsicologico', '/Scripts/apps/DiagnosticoPsicologico/views/Row.html');

    Templates.load('respuesta1', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta1.html');
    Templates.load('respuesta2', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta2.html');
    Templates.load('respuesta3', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta3.html');
    Templates.load('respuesta4', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta4.html');
    Templates.load('respuesta5', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta5.html');
    Templates.load('respuesta6', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta6.html');
    Templates.load('respuesta7', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta7.html');
    Templates.load('respuesta8', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta8.html');
    Templates.load('respuesta9', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta9.html');
    Templates.load('respuesta10', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta10.html');
    Templates.load('respuesta11', '/Scripts/apps/DiagnosticoPsicologico/views/Respuesta11.html');
}

